### DISCUSSION TITLE
Overly click through transparent window

### DISCUSSION THREAD

**Discussion by MajdAlbaho** [AvaloniaUI/Avalonia#13827] [General]
I did not find any solution for creating overlay transparent click-through window, I wanna make a window and pass mouse click through it over operating system apps

In WPF I'm able to implement this by setting the background = Transparent, WindowState = None, AllowsTransparent = True
But those settings are not exist in Avalonia

Any suggestions ?

**Comment by thevortexcloud** [comment]
As far as I am aware: There is no built in way to do this. You can however do it via various platform specific native APIs. On windows this is fairly straightforwards (you need to set `WS_EX_TRANSPARENT` on the window). On Linux (specifically X11) it requires the use of various obscure and not very well documented _optional_ [X11 APIs](https://en.wikipedia.org/wiki/Shape_extension) (although most distros ship with them anyway). The X11 APIs are more flexible than Window's though. They essentially let you cut arbitrary holes in the window. But they also require more steps to use since X11 is a server protocol so you need to connect to the server and what not.

For every other platform I can't say. I believe it's not possible at all on Wayland due to security concerns. But I could be wrong.

**Reply by MajdAlbaho** [reply]
You're right, there is no build in functions and it's not possible on Wayland due to high level security, just to mention that each platform has its implementation via native APIs

**Reply by infinitecubes0-ai** [reply]
wait is there a way now? Been 2 years since you posted this.

**Reply by thevortexcloud** [reply]
As far as I aware: Nothing has changed. Wayland is a fundamentally different approach than X11. One of Wayland's big goals was to fix all the security problems X11 had. Even things like global shortcuts took ages to be implemented due to fear of people making key loggers. Furthermore, Wayland itself is just a protocol. There are _many_ implementations of the protocol which all support different things and behave differently.

That said: It _could_ probably be done with a compositor extension/custom compositor, but I don't know of any that exist that do this. KDE for example has some extensions to allow for fake user input https://wayland.app/protocols/. But a custom compositor/extension is not particularly useful outside of an embedded environment.

Anyway: It has been a number of years since I did any in-depth research on this. So perhaps something has changed.

**Reply by infinitecubes0-ai** [reply]
> That said: It _could_ probably be done with a compositor extension/custom compositor, but I don't know of any that exist that do this. KDE for example has some extensions to allow for fake user input https://wayland.app/protocols/. But a custom compositor/extension is not particularly useful outside of an embedded environment.
> 


I was testing on Hyprland with a steam game (similar setup to bongo cat on steam, but I have early demo to it (desktropolis)).  Like bongo cat, the background is completely black.  I did find some scripts (correction: ai found some scripts) to make part (or all) of the window click-through.  However, it only works for Xwayland windows.  Apparently Steam and Steam games are xwayland.  I don’t think anything else really is.

anyway basically when the script was active then inputs would pass to the next xwayland window regardless of what desktop they’re in.  So basically my inputs would pass to my steam library page, regardless of what was behind Desktropolis.

So there ARE ways to make it work for xwayland.  But for wayland windows, it doesn’t.

**Reply by thevortexcloud** [reply]
>  Apparently Steam and Steam games are xwayland. I don’t think anything else really is.

Actually most things are still XWayland (Including Avalonia). There has only been a major concerted effort for native Wayland in the last ~5 years. A lot of old programs will likely never get updated for native Wayland. Even the ones that are can often be changed back to X11. EG all modern QT programs can be easily changed by setting some environment variables.

But everything is slowly moving over. Some things will never move because Wayland will likely never support the features they need. EG a lot of people _really_ want to be able to position their windows manually with specific coordinates. Which Wayland has (understandably) categorically refused to support.

Also: Some games are actually native wayland. If they use SDL, Wayland can also often be enabled/disabled by the user. SDL3 enables it by default.

https://wiki.libsdl.org/SDL3/README-wayland
